include/dgl/array.h

```diff
@@ -239,11 +242,38 @@ struct CSRMatrix {
   /*! \brief default constructor */
   CSRMatrix() = default;
   /*! \brief constructor */
-  CSRMatrix(int64_t nrows, int64_t ncols,
-            IdArray parr, IdArray iarr, IdArray darr = IdArray(),
-            bool sorted_flag = false)
-    : num_rows(nrows), num_cols(ncols), indptr(parr), indices(iarr),
-      data(darr), sorted(sorted_flag) {}
+  CSRMatrix(int64_t nrows, int64_t ncols, IdArray parr, IdArray iarr,
+            IdArray darr = IdArray(), bool sorted_flag = false)
+      : num_rows(nrows),
+        num_cols(ncols),
+        indptr(parr),
+        indices(iarr),
+        data(darr),
+        sorted(sorted_flag) {}
+
+  bool Load(dmlc::Stream* fs) {
+    uint64_t magicNum;
+    CHECK(fs->Read(&magicNum)) << "Invalid Magic Number";
+    CHECK_EQ(magicNum, kDGLSerialize_AtenCsrMatrixMagic)
+        << "Invalid CSRMatrix Data";
+    CHECK(fs->Read(&num_cols)) << "Invalid num_cols";
+    CHECK(fs->Read(&num_rows)) << "Invalid num_rows";
+    CHECK(fs->Read(&indptr)) << "Invalid indptr";
+    CHECK(fs->Read(&indices)) << "Invalid indices";
+    CHECK(fs->Read(&data)) << "Invalid data";
+    CHECK(fs->Read(&sorted)) << "Invalid sorted";
+    return true;
+  }
+
+  void Save(dmlc::Stream* fs) const {
+    fs->Write(kDGLSerialize_AtenCsrMatrixMagic);
+    fs->Write(num_cols);
+    fs->Write(num_rows);
+    fs->Write(indptr);
+    fs->Write(indices);
+    fs->Write(data);
```

Comment:
Just a question. Will this correctly handle the case of when `data` is initialized with default constructor? Note that default constructor gives undefined NDArray, whose internal pointer is null.

Reply:
It creates a new empty NDArray and set it to the current object
https://github.com/dmlc/dgl/blob/master/include/dgl/runtime/ndarray.h#L444-L464

Reply:
For `NDArray` directly created by default constructor, its internal `data_` pointer is null, so this line will fail:
https://github.com/dmlc/dgl/blob/c3a33407fe8b4938156427c701f0386dfc1a38de/include/dgl/runtime/ndarray.h#L416-L418
https://github.com/dmlc/dgl/blob/c3a33407fe8b4938156427c701f0386dfc1a38de/include/dgl/runtime/ndarray.h#L360-L362
I think the solution is to avoid such thing, which means we should not have the default argument of `data` to be `IdArray()`. I guess we could define a special ndarray to represent empty, which is equal to `NDArray({}, DLDataType{kDLInt, 32, 1}, DLContext{kDLCPU, 0})`. What do you think?

Reply:
cc @BarclayII here too. We might need to do a pass on all the CSR/COO operators to make sure that the data array, when omitted, is correctly set to a defined empty array.

Reply:
Agree.  An undefined NDArray is probably a bad idea anyway.

In this case, it's probably better to have a static function
```C++
static NDArray NDArray::Empty(DLDataType dtype, DLContext ctx)
```

Although we already have a function named `Empty` for different purpose.